cw/cw305/capture.py

```diff
@@ -330,6 +330,124 @@ def capture_aes_fvsr_key(ot):
+            for ii in range(scope.num_segments_actual):
+                if sample_fixed:
+                    key = np.asarray(key_fixed)
+                else:
+                    key = np.asarray(ktp.next()[1])
+                plaintext = np.asarray(ktp.next()[1])
+                ciphertext = np.asarray(scared.aes.base.encrypt(plaintext, key))
+                plaintexts.append(plaintext)
+                keys.append(key)
+                ciphertexts.append(ciphertext)
+                sample_fixed = plaintext[0] & 0x1
+
+            # dummy operation for synching PRNGs with the firmware
+            for ii in range(scope.num_segments_actual):
+                _ = np.asarray(ktp.next()[1])
+
+            # Check the ciphertext_total of this batch to make sure we are in sync.
+            actual_last_ciphertext = ot.target.simpleserial_read("r", 16, ack=False)
+            expected_last_ciphertext = ciphertext
+            assert actual_last_ciphertext == expected_last_ciphertext, (
+                f"Incorrect encryption result!\n"
+                f"actual: {actual_last_ciphertext}\n"
+                f"expected: {expected_last_ciphertext}"
+            )
+
+            # Make sure to allocate sufficient memory for the storage segment array during the
+            # first resize operation. By default, the ChipWhisperer API starts every new segment
+            # with 1 trace and then increases it on demand by 25 traces at a time. This results in
+            # frequent array resizing and decreasing capture rate.
+            # See addWave() in chipwhisperer/common/traces/_base.py.
+            if project.traces.cur_seg.tracehint < project.traces.seg_len:
+                project.traces.cur_seg.setTraceHint(project.traces.seg_len)
+            # Only keep the latest two trace storage segments enabled. By default the ChipWhisperer
+            # API keeps all segments enabled and after appending a new trace, the trace ranges are
+            # updated for all segments. This leads to a decreasing capture rate after time.
+            # See:
+            # - _updateRanges() in chipwhisperer/common/api/TraceManager.py.
+            # - https://github.com/newaetech/chipwhisperer/issues/344
+            if num_segments_storage != len(project.segments):
+                if num_segments_storage >= 2:
+                    project.traces.tm.setTraceSegmentStatus(num_segments_storage - 2, False)
+                num_segments_storage = len(project.segments)
+
+            # Add traces of this batch to the project.
+            for wave, plaintext, ciphertext, key in zip(waves, plaintexts, ciphertexts, keys):
+                project.traces.append(
+                    cw.common.traces.Trace(wave, plaintext, ciphertext, key),
+                    dtype=np.uint16
+                )
+
+            # Update the loop variable and the progress bar.
+            rem_num_traces -= scope.num_segments
+            pbar.update(scope.num_segments)
```

Comment:
This code is exactly the same as for `capture_aes_random_batch()` does it make sense to share it by creating functions?

Reply:
I was thinking the same. What would be better: In the same PR or in a seperate PR?